FlinkRuntimeException: Failed to deserialize value

<!--
    possible labels: bug,datastream-api,table-api
-->
## Description

Flink 2.1.0, Java17, Clickhouse 25.8.1 on Ubuntu24

### Steps to reproduce
1. Tried to inject 3 billion records from Entities that are instantiated from CSV files. No errors in the conversion.
2. Used 10 TaskManagers, `env.setParallelism(40)`
3. At some point I get this error, in my case after ~200 million Entities are injected successfully
### Error Log or Exception StackTrace

```
2025-09-09 14:21:46,187 WARN  org.apache.flink.runtime.taskmanager.Task                    [] - Source: Gzipped CSV Source -> Map -> Filter -> Sink: Writer (17/40)#81 (e01de927fad51df3c0211700c3d0219a_cbc357ccb763df2852fee8c4fc7d55f2_16_81) switched from INITIALIZING to FAILED with failure cause:
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
	at org.apache.flink.streaming.api.operators.util.SimpleVersionedListState$DeserializingIterator.next(SimpleVersionedListState.java:140) ~[flink-dist-2.1.0.jar:2.1.0]
	at java.base/java.util.Iterator.forEachRemaining(Unknown Source) ~[?:?]
	at org.apache.flink.util.CollectionUtil.iterableToList(CollectionUtil.java:119) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.operators.sink.StatefulSinkWriterStateHandler.createWriter(StatefulSinkWriterStateHandler.java:103) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.operators.sink.SinkWriterOperator.initializeState(SinkWriterOperator.java:171) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.initializeOperatorState(StreamOperatorStateHandler.java:142) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.initializeState(AbstractStreamOperator.java:304) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.initializeStateAndOpenOperators(RegularOperatorChain.java:106) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restoreStateAndGates(StreamTask.java:858) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$restoreInternal$5(StreamTask.java:812) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.call(StreamTaskActionExecutor.java:55) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restoreInternal(StreamTask.java:812) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restore(StreamTask.java:771) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:963) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:932) [flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:756) [flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:568) [flink-dist-2.1.0.jar:2.1.0]
	at java.base/java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.io.IOException: Unsupported version: 1123
	at org.apache.flink.connector.clickhouse.sink.ClickHouseAsyncSinkSerializer.deserializeRequestFromStream(ClickHouseAsyncSinkSerializer.java:41) ~[blob_p-ea3ad9f3732e1fa10aca39e3bd3cecfd6494fb2b-15c484d7f218d531026537aac0c41986:?]
	at org.apache.flink.connector.clickhouse.sink.ClickHouseAsyncSinkSerializer.deserializeRequestFromStream(ClickHouseAsyncSinkSerializer.java:12) ~[blob_p-ea3ad9f3732e1fa10aca39e3bd3cecfd6494fb2b-15c484d7f218d531026537aac0c41986:?]
	at org.apache.flink.connector.base.sink.writer.AsyncSinkWriterStateSerializer.deserialize(AsyncSinkWriterStateSerializer.java:81) ~[flink-connector-files-2.1.0.jar:2.1.0]
	at org.apache.flink.connector.base.sink.writer.AsyncSinkWriterStateSerializer.deserialize(AsyncSinkWriterStateSerializer.java:39) ~[flink-connector-files-2.1.0.jar:2.1.0]
	at org.apache.flink.core.io.SimpleVersionedSerialization.readVersionAndDeSerialize(SimpleVersionedSerialization.java:227) ~[flink-dist-2.1.0.jar:2.1.0]
	at org.apache.flink.streaming.api.operators.util.SimpleVersionedListState$DeserializingIterator.next(SimpleVersionedListState.java:138) ~[flink-dist-2.1.0.jar:2.1.0]
	... 17 more
```

```
$ kube logs  flink10-taskmanager-6 | grep Exception | uniq
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1327
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1361
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1123
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1224
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1257
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1058
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1176
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1139
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
Caused by: java.io.IOException: Unsupported version: 1256
org.apache.flink.util.FlinkRuntimeException: Failed to deserialize value
...
```

### Expected Behaviour

### Code Example

Using `POJOConvertor<T>` only with defaultsSupport = false, hasDefault = false. Respected the exact order of columns in the definition.
```java
Serialize.writeInt32(out, input.field1(), false, false, ClickHouseDataType.Int32, false, "field1");
Serialize.writeString(out, input.field2(), false, false, ClickHouseDataType.String, false, "field2");
Serialize.writeTimeDate(out, input.field3(), false, false, ClickHouseDataType.DateTime, false, "field3");
...
```

### Configuration

#### Client Configuration
```java
var clickHouseClientConfig = new ClickHouseClientConfig(
            // internal IP of ClickHouse instance
            url,
            "default",
            "_____________________________",
            "alerts",
            "alerts"
        );
        clickHouseClientConfig.setSupportDefault(false);

        // convertor & sink
        var entityConvertor = new EntityJsonOutputStreamConvertor();
        var convertorEntity = new ClickHouseConvertor<>(EntityJsonRow.class, entityConvertor);
        var clickhouseSink = new ClickHouseAsyncSink<>(
            convertorEntity,
            MAX_BATCH_SIZE,
            MAX_IN_FLIGHT_REQUESTS,
            MAX_BUFFERED_REQUESTS,
            MAX_BATCH_SIZE_IN_BYTES,
            MAX_TIME_IN_BUFFER_MS,
            MAX_RECORD_SIZE_IN_BYTES,
            clickHouseClientConfig
        );
        events.sinkTo(clickhouseSink);
```

#### Environment
* [ ] Cloud
* Connector version: master, changed Flink libraries to point to 2.1.0, Java17
* Language version: 17
* OS: Ubuntu 24

#### ClickHouse Server
* ClickHouse Server version: 25.8.1
* ClickHouse Server non-default settings, if any:
* `CREATE TABLE` statements for tables involved:
* Sample data for all these tables, use [clickhouse-obfuscator](https://github.com/ClickHouse/ClickHouse/blob/master/programs/obfuscator/Obfuscator.cpp#L42-L80) if necessary

--- a/flink-connector-clickhouse-1.17/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
+++ b/flink-connector-clickhouse-1.17/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
@@ -20,6 +20,7 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
             dataOutputStream.writeInt(bytes.length);
             dataOutputStream.write(bytes);
         } else {
+            dataOutputStream.writeInt(V1);
             dataOutputStream.writeInt(-1);
         }
 
@@ -41,10 +42,10 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
             if (version == V1) {
                 return deserializeV1(dataInputStream);
             } else {
-                throw new IOException("Unsupported version " + version);
+                throw new IOException("Unsupported serialization version: " + version);
             }
         } else {
-            throw new IOException("Request size: " + requestSize);
+            throw new IOException("Invalid request size: " + requestSize);
         }
     }
 

--- a/flink-connector-clickhouse-1.17/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
+++ b/flink-connector-clickhouse-1.17/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
@@ -44,7 +44,7 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
                 throw new IOException("Unsupported version " + version);
             }
         } else {
-            throw new IOException("request size < 0");
+            throw new IOException("Request size: " + requestSize);
         }
     }
 

--- a/flink-connector-clickhouse-1.17/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
+++ b/flink-connector-clickhouse-1.17/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
@@ -16,6 +16,7 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
     protected void serializeRequestToStream(ClickHousePayload clickHousePayload, DataOutputStream dataOutputStream) throws IOException {
         byte[] bytes = clickHousePayload.getPayload();
         if (bytes != null) {
+            dataOutputStream.writeInt(V1);
             dataOutputStream.writeInt(bytes.length);
             dataOutputStream.write(bytes);
         } else {
@@ -34,11 +35,16 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
     }
 
     @Override
-    protected ClickHousePayload deserializeRequestFromStream(long version, DataInputStream dataInputStream) throws IOException {
-        if (version == V1) {
-            return deserializeV1(dataInputStream);
+    protected ClickHousePayload deserializeRequestFromStream(long requestSize, DataInputStream dataInputStream) throws IOException {
+        if  (requestSize > 0) {
+            int version = dataInputStream.readInt();
+            if (version == V1) {
+                return deserializeV1(dataInputStream);
+            } else {
+                throw new IOException("Unsupported version " + version);
+            }
         } else {
-            throw new IOException("Unsupported version: " + version);
+            throw new IOException("request size < 0");
         }
     }
 

--- a/flink-connector-clickhouse-1.17/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
+++ b/flink-connector-clickhouse-1.17/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
@@ -1,0 +1,29 @@
+package org.apache.flink.connector.clickhouse.sink;
+
+import org.apache.flink.connector.clickhouse.data.ClickHousePayload;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+
+public class ClickHouseSinkStateTests {
+
+    @Test
+    void SerializerTest() throws Exception {
+        byte[] data = {'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd'};
+        ClickHousePayload clickHousePayload = new ClickHousePayload(data);
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream dos = new DataOutputStream(baos);
+        ClickHouseAsyncSinkSerializer serializer = new ClickHouseAsyncSinkSerializer();
+        serializer.serializeRequestToStream(clickHousePayload, dos);
+        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
+
+        ClickHousePayload clickHousePayload1 = serializer.deserializeRequestFromStream(dos.size(), dis);
+        Assertions.assertEquals(clickHousePayload.getPayloadLength(), clickHousePayload1.getPayloadLength());
+        Assertions.assertArrayEquals(clickHousePayload.getPayload(), clickHousePayload1.getPayload());
+    }
+}

--- a/flink-connector-clickhouse-1.17/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
+++ b/flink-connector-clickhouse-1.17/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
@@ -4,15 +4,12 @@ import org.apache.flink.connector.clickhouse.data.ClickHousePayload;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
-import java.io.DataInputStream;
-import java.io.DataOutputStream;
+import java.io.*;
 
 public class ClickHouseSinkStateTests {
 
     @Test
-    void SerializerTest() throws Exception {
+    void testSerializeAndDeserializePayload() throws Exception {
         byte[] data = {'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd'};
         ClickHousePayload clickHousePayload = new ClickHousePayload(data);
 
@@ -25,5 +22,37 @@ public class ClickHouseSinkStateTests {
         ClickHousePayload clickHousePayload1 = serializer.deserializeRequestFromStream(dos.size(), dis);
         Assertions.assertEquals(clickHousePayload.getPayloadLength(), clickHousePayload1.getPayloadLength());
         Assertions.assertArrayEquals(clickHousePayload.getPayload(), clickHousePayload1.getPayload());
+    }
+
+    @Test
+    void testSerializeAndDeserializeEmptyPayload() throws Exception {
+        ClickHousePayload clickHousePayload = new ClickHousePayload(null);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream dos = new DataOutputStream(baos);
+        ClickHouseAsyncSinkSerializer serializer = new ClickHouseAsyncSinkSerializer();
+        serializer.serializeRequestToStream(clickHousePayload, dos);
+        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
+        ClickHousePayload clickHousePayload1 = serializer.deserializeRequestFromStream(dos.size(), dis);
+        Assertions.assertEquals(clickHousePayload.getPayloadLength(), clickHousePayload1.getPayloadLength());
+        Assertions.assertArrayEquals(clickHousePayload.getPayload(), clickHousePayload1.getPayload());
+    }
+
+    @Test
+    void testDeserializePayloadWithUnsuportedVersion() throws IOException {
+        byte[] data = {'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd'};
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream dos = new DataOutputStream(baos);
+        DataOutputStream dataOutputStream = new DataOutputStream(baos);
+        int V2 = 2;
+        dataOutputStream.writeInt(V2);
+        dataOutputStream.writeInt(data.length);
+        dataOutputStream.write(data);
+        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
+
+        ClickHouseAsyncSinkSerializer serializer = new ClickHouseAsyncSinkSerializer();
+        Exception exception = Assertions.assertThrows(IOException.class, () -> {
+            serializer.deserializeRequestFromStream(dataOutputStream.size(), dis);
+        });
+        Assertions.assertEquals("Unsupported serialization version: 2", exception.getMessage());
     }
 }

--- a/flink-connector-clickhouse-1.17/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkTests.java
+++ b/flink-connector-clickhouse-1.17/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkTests.java
@@ -45,6 +45,7 @@ public class ClickHouseSinkTests extends FlinkClusterTests {
     static final long MAX_RECORD_SIZE_IN_BYTES = 1000;
 
     static final int STREAM_PARALLELISM = 5;
+    static final int NUMBER_OF_RETRIES = 10;
 
     private String createSimplePOJOTableSQL(String database, String tableName, int parts_to_throw_insert) {
         String createTable = createSimplePOJOTableSQL(database, tableName);
@@ -511,7 +512,7 @@ public class ClickHouseSinkTests extends FlinkClusterTests {
         env.setParallelism(STREAM_PARALLELISM);
 
         ClickHouseClientConfig clickHouseClientConfig = new ClickHouseClientConfig(getServerURL(), getUsername(), getPassword(), getDatabase(), tableName);
-        clickHouseClientConfig.setNumberOfRetries(10);
+        clickHouseClientConfig.setNumberOfRetries(NUMBER_OF_RETRIES);
         clickHouseClientConfig.setSupportDefault(simpleTableSchema.hasDefaults());
 
         ElementConverter<SimplePOJO, ClickHousePayload> convertorCovid = new ClickHouseConvertor<>(SimplePOJO.class, simplePOJOConvertor);

--- a/flink-connector-clickhouse-2.0.0/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
+++ b/flink-connector-clickhouse-2.0.0/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
@@ -20,6 +20,7 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
             dataOutputStream.writeInt(bytes.length);
             dataOutputStream.write(bytes);
         } else {
+            dataOutputStream.writeInt(V1);
             dataOutputStream.writeInt(-1);
         }
 
@@ -41,10 +42,10 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
             if (version == V1) {
                 return deserializeV1(dataInputStream);
             } else {
-                throw new IOException("Unsupported version " + version);
+                throw new IOException("Unsupported serialization version: " + version);
             }
         } else {
-            throw new IOException("Request size: " + requestSize);
+            throw new IOException("Invalid request size: " + requestSize);
         }
     }
 

--- a/flink-connector-clickhouse-2.0.0/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
+++ b/flink-connector-clickhouse-2.0.0/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
@@ -44,7 +44,7 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
                 throw new IOException("Unsupported version " + version);
             }
         } else {
-            throw new IOException("request size < 0");
+            throw new IOException("Request size: " + requestSize);
         }
     }
 

--- a/flink-connector-clickhouse-2.0.0/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
+++ b/flink-connector-clickhouse-2.0.0/src/main/java/org/apache/flink/connector/clickhouse/sink/ClickHouseAsyncSinkSerializer.java
@@ -16,6 +16,7 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
     protected void serializeRequestToStream(ClickHousePayload clickHousePayload, DataOutputStream dataOutputStream) throws IOException {
         byte[] bytes = clickHousePayload.getPayload();
         if (bytes != null) {
+            dataOutputStream.writeInt(V1);
             dataOutputStream.writeInt(bytes.length);
             dataOutputStream.write(bytes);
         } else {
@@ -34,11 +35,16 @@ public class ClickHouseAsyncSinkSerializer extends AsyncSinkWriterStateSerialize
     }
 
     @Override
-    protected ClickHousePayload deserializeRequestFromStream(long version, DataInputStream dataInputStream) throws IOException {
-        if (version == V1) {
-            return deserializeV1(dataInputStream);
+    protected ClickHousePayload deserializeRequestFromStream(long requestSize, DataInputStream dataInputStream) throws IOException {
+        if  (requestSize > 0) {
+            int version = dataInputStream.readInt();
+            if (version == V1) {
+                return deserializeV1(dataInputStream);
+            } else {
+                throw new IOException("Unsupported version " + version);
+            }
         } else {
-            throw new IOException("Unsupported version: " + version);
+            throw new IOException("request size < 0");
         }
     }
 

--- a/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
+++ b/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
@@ -1,0 +1,29 @@
+package org.apache.flink.connector.clickhouse.sink;
+
+import org.apache.flink.connector.clickhouse.data.ClickHousePayload;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+
+public class ClickHouseSinkStateTests {
+
+    @Test
+    void SerializerTest() throws Exception {
+        byte[] data = {'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd'};
+        ClickHousePayload clickHousePayload = new ClickHousePayload(data);
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream dos = new DataOutputStream(baos);
+        ClickHouseAsyncSinkSerializer serializer = new ClickHouseAsyncSinkSerializer();
+        serializer.serializeRequestToStream(clickHousePayload, dos);
+        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
+
+        ClickHousePayload clickHousePayload1 = serializer.deserializeRequestFromStream(dos.size(), dis);
+        Assertions.assertEquals(clickHousePayload.getPayloadLength(), clickHousePayload1.getPayloadLength());
+        Assertions.assertArrayEquals(clickHousePayload.getPayload(), clickHousePayload1.getPayload());
+    }
+}

--- a/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
+++ b/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkStateTests.java
@@ -4,15 +4,12 @@ import org.apache.flink.connector.clickhouse.data.ClickHousePayload;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
-import java.io.DataInputStream;
-import java.io.DataOutputStream;
+import java.io.*;
 
 public class ClickHouseSinkStateTests {
 
     @Test
-    void SerializerTest() throws Exception {
+    void testSerializeAndDeserializePayload() throws Exception {
         byte[] data = {'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd'};
         ClickHousePayload clickHousePayload = new ClickHousePayload(data);
 
@@ -21,9 +18,40 @@ public class ClickHouseSinkStateTests {
         ClickHouseAsyncSinkSerializer serializer = new ClickHouseAsyncSinkSerializer();
         serializer.serializeRequestToStream(clickHousePayload, dos);
         DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
-
         ClickHousePayload clickHousePayload1 = serializer.deserializeRequestFromStream(dos.size(), dis);
         Assertions.assertEquals(clickHousePayload.getPayloadLength(), clickHousePayload1.getPayloadLength());
         Assertions.assertArrayEquals(clickHousePayload.getPayload(), clickHousePayload1.getPayload());
+    }
+
+    @Test
+    void testSerializeAndDeserializeEmptyPayload() throws Exception {
+        ClickHousePayload clickHousePayload = new ClickHousePayload(null);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream dos = new DataOutputStream(baos);
+        ClickHouseAsyncSinkSerializer serializer = new ClickHouseAsyncSinkSerializer();
+        serializer.serializeRequestToStream(clickHousePayload, dos);
+        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
+        ClickHousePayload clickHousePayload1 = serializer.deserializeRequestFromStream(dos.size(), dis);
+        Assertions.assertEquals(clickHousePayload.getPayloadLength(), clickHousePayload1.getPayloadLength());
+        Assertions.assertArrayEquals(clickHousePayload.getPayload(), clickHousePayload1.getPayload());
+    }
+
+    @Test
+    void testDeserializePayloadWithUnsuportedVersion() throws IOException {
+        byte[] data = {'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd'};
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream dos = new DataOutputStream(baos);
+        DataOutputStream dataOutputStream = new DataOutputStream(baos);
+        int V2 = 2;
+        dataOutputStream.writeInt(V2);
+        dataOutputStream.writeInt(data.length);
+        dataOutputStream.write(data);
+        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
+
+        ClickHouseAsyncSinkSerializer serializer = new ClickHouseAsyncSinkSerializer();
+        Exception exception = Assertions.assertThrows(IOException.class, () -> {
+            serializer.deserializeRequestFromStream(dataOutputStream.size(), dis);
+        });
+        Assertions.assertEquals("Unsupported serialization version: 2", exception.getMessage());
     }
 }

--- a/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkTests.java
+++ b/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/clickhouse/sink/ClickHouseSinkTests.java
@@ -45,6 +45,7 @@ public class ClickHouseSinkTests extends FlinkClusterTests {
     static final long MAX_RECORD_SIZE_IN_BYTES = 1000;
 
     static final int STREAM_PARALLELISM = 5;
+    static final int NUMBER_OF_RETRIES = 10;
 
     private String createSimplePOJOTableSQL(String database, String tableName, int parts_to_throw_insert) {
         String createTable = createSimplePOJOTableSQL(database, tableName);
@@ -508,7 +509,7 @@ public class ClickHouseSinkTests extends FlinkClusterTests {
         env.setParallelism(STREAM_PARALLELISM);
 
         ClickHouseClientConfig clickHouseClientConfig = new ClickHouseClientConfig(getServerURL(), getUsername(), getPassword(), getDatabase(), tableName);
-        clickHouseClientConfig.setNumberOfRetries(10);
+        clickHouseClientConfig.setNumberOfRetries(NUMBER_OF_RETRIES);
         clickHouseClientConfig.setSupportDefault(simpleTableSchema.hasDefaults());
 
         ElementConverter<SimplePOJO, ClickHousePayload> convertorCovid = new ClickHouseConvertor<>(SimplePOJO.class, simplePOJOConvertor);

--- a/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/test/embedded/flink/EmbeddedFlinkClusterForTests.java
+++ b/flink-connector-clickhouse-2.0.0/src/test/java/org/apache/flink/connector/test/embedded/flink/EmbeddedFlinkClusterForTests.java
@@ -21,6 +21,10 @@ public class EmbeddedFlinkClusterForTests {
 
     public static void setUp() throws Exception {
         Configuration config = new Configuration();
+        setUp(config);
+    }
+
+    public static void setUp(Configuration config) throws Exception {
         config.set(RestOptions.PORT, REST_PORT); // web UI port (optional)
         config.set(TaskManagerOptions.NUM_TASK_SLOTS, NUM_TASK_SLOTS);
         flinkCluster = new MiniClusterWithClientResource(
@@ -31,6 +35,7 @@ public class EmbeddedFlinkClusterForTests {
                         .build());
         flinkCluster.before();
     }
+
     public static void tearDown() {
         if (flinkCluster != null) {
             flinkCluster.after();

--- a/flink-connector-clickhouse-base/src/main/java/org/apache/flink/connector/clickhouse/data/ClickHousePayload.java
+++ b/flink-connector-clickhouse-base/src/main/java/org/apache/flink/connector/clickhouse/data/ClickHousePayload.java
@@ -15,7 +15,10 @@ public class ClickHousePayload implements Serializable {
         this.payload = payload;
     }
     public byte[] getPayload() { return payload; }
-    public int getPayloadLength() { return payload.length; }
+    public int getPayloadLength() {
+        if (payload == null) return -1;
+        return payload.length;
+    }
     public int getAttemptCount() { return attemptCount; }
     public void incrementAttempts() { attemptCount++; }
 }

